test/integration/puppetlabs/services/jruby/jruby_locking_test.clj

```diff
@@ -28,14 +47,54 @@
         (is (not (.isWriteLocked lock))))
 
       (testing "with-lock macro releases lock even if body throws exception"
-        (is (thrown? RuntimeException
-                     (jruby-service/with-lock jruby-service
+        (is (thrown? IllegalStateException
+                     (jruby-service/with-lock jruby-service :with-lock-exception-test
                        (is (.isWriteLocked lock))
-                       (throw (RuntimeException. "exception")))))
+                       (throw (IllegalStateException. "exception")))))
         (is (not (.isWriteLocked lock)))))))
 
+(deftest ^:integration with-lock-event-notification-test
+  (testing "locking sends event notifications"
+    (let [events (atom [])
+          callback (fn [{:keys [type] :as event}]
+                     (swap! events conj type))
+          event-service (tk/service [[:JRubyPuppetService register-event-handler]]
+                                    (init [this context]
+                                          (register-event-handler callback)
+                                          context))]
+      (tk-bootstrap/with-app-with-config
+        app
+        [jruby-service/jruby-puppet-pooled-service
+         profiler/puppet-profiler-service
+         event-service]
+        (jruby-service-test-config 1)
+        (let [jruby-service (tk-app/get-service app :JRubyPuppetService)]
+
+          (testing "locking events trigger event notifications"
+            (jruby-service/with-jruby-puppet
+              jruby-puppet
+              jruby-service
+              :with-lock-events-test
+              (testing "borrowing a jruby triggers 'requested'/'borrow' events"
+                (is (= [:instance-requested :instance-borrowed] @events))))
+            (testing "returning a jruby triggers 'returned' event"
+              (is (= [:instance-requested :instance-borrowed :instance-returned] @events)))
+            (jruby-service/with-lock
+              jruby-service
+              :with-lock-events-test
+              (testing "acquiring a lock triggers 'lock-requested'/'lock-acquired' events"
+                (is (= [:instance-requested :instance-borrowed :instance-returned
+                        :lock-requested :lock-acquired] @events)))))
+          (testing "releasing the lock triggers 'lock-released' event"
+            (is (= [:instance-requested :instance-borrowed :instance-returned
+                    :lock-requested :lock-acquired :lock-released] @events))))))))
```

Comment:
aaaand now I understand why you wrote the File Sync version of this test using an atom.

![](http://vignette1.wikia.nocookie.net/fbandcc/images/b/b9/The_less_you_know.JPG/revision/latest/scale-to-width-down/640?cb=20120223192136)